aten/src/ATen/native/sparse/SparseCsrTensor.cpp

```diff
@@ -798,16 +800,72 @@ Tensor select_sparse_csr(const Tensor& self, int64_t dim, int64_t index) {
         options.pinned_memory_opt());
   } else if (dim < n_batch + 2) {
     // Selecting sparse dimension
-    TORCH_CHECK(
-        self.layout() == kSparseCsr || self.layout() == kSparseCsc,
-        "select(): selecting non-batch dimensions is currently only supported for non-blocked sparse compressed layouts tensors.");
     TORCH_CHECK(
         n_batch == 0,
-        "select(): selecting rows or columns is not implemented for batched sparse compressed tensors.")
-    // Converting to COO and calling select is slightly slower than operating
-    // on the CSR indices directly for constructing a COO vector, however
-    // current version is more readable and easier to understand.
-    return self.to_sparse().select(dim, index);
+        "select(): selecting sparse dimensions is not implemented for batched sparse compressed tensors.")
+    TORCH_INTERNAL_ASSERT(dim == 0 || dim == 1);
+    DimVector blocksize{1, 1};
+    DimVector output_shape = DimVector(self.sizes().slice(1, self.dim()-1));
+    if (dim == 1) {
+      output_shape[0] = self.size(0);
+    }
+    auto indices_options = compressed_indices.options();
+    return AT_DISPATCH_PLAIN_SPARSE_COMPRESSED_LAYOUTS(self.layout(), "select()",
+        [&]() {
+          /* The following is equivalent to
+             self.to_sparse().select(dim, index) but can be several
+             orders of magnitude faster depending on the input shape
+             and select dim.
+          */
+          Tensor dim_indices;
+          Tensor indices;
+          if ((the_layout == kSparseCsr && dim == 0) || (the_layout == kSparseCsc && dim == 1)) {
+            auto start_end = compressed_indices.slice(0, index, index + 2, 1).cpu();
+            AT_DISPATCH_INDEX_TYPES(start_end.scalar_type(), "select()", [&]() {
+              auto start_end_accessor = start_end.accessor<index_t, 1>();
+              dim_indices = at::arange(start_end_accessor[0], start_end_accessor[1], indices_options);
+            });
+            indices = plain_indices;
+          } else {
+            dim_indices = at::where(plain_indices.eq(index))[0];
+            indices = at::_convert_indices_from_csr_to_coo(compressed_indices, plain_indices).select(0, 0);
+          }
+          Tensor values = self.values().index_select(0, dim_indices);
+          return at::_sparse_coo_tensor_unsafe(indices.index_select(0, dim_indices).view({1, -1}).to(kLong), values, output_shape);
```

Comment:
I have a suspicion that this logic could get simplified with something like (CSR in mind):
```python
crow_slice = crow_indices.narrow(dim, index, 2).cpu()
crow_start = crow_slice[0]
crow_end = crow_slice[1]
new_values = values.slice(dim, crow_start, crow_end)
coo_index = torch.stack((torch.tensor(idx, ...).expand(...), col_indices.narrow(dim, crow_start, crow_end)))
return torch._sparse_coo_tensor_unsafe(coo_index, new_values, ...)
```

Reply:
If `n_batch != 0`, I think we can still implement it similarly but with `index` instead of `slice`.

Reply:
yes, something like this would replace the `arange` part of the code above that would lead to some more speed-up. Note that this will correspond to the `a.select(0, 0)` case which is already quite performant (compared to the current state).

A more interesting question is if the `where` part of the code (that corresponds to the `a.select(1, 0)` case) could be made faster.

Reply:
Oh, sorry, we do not even need `stack` since `select` reduces the dim.

Reply:
Also note that `slice` will always create a view on `values`, while `index_select` will create a copy. I guess `select` is supposed to be a view...

Reply:
The suggestion increased the performance of `csr.select(0, 0)` by 1.5-2 times.

The usage of `index_select` cannot be avoided for `dim=1` case, so `compressed_sparse.select(1, ..)` cannot be a view, in general.